### ISSUE TITLE
vaadinBuildFrontend irregularly failing with ZipException: ZipFile invalid LOC header (bad signature)

### ISSUE THREAD

**Issue by skiedrowski** [vaadin/flow#15458]
**Desktop (please complete the following information):**
 - OS: macOS 12.x, macOS 13.x, Ubuntu 22.4, Windows 11
 - JDK version: 11 (temurin + zulu tested)
 - Gradle version: 7.5.1
 - Vaadin Plugin version: 23.1.17

**Describe the bug**
Running `./gradlew vaadinBuildFrontend` sometimes fails with `ZipException: ZipFile invalid LOC header (bad signature)` ([Stacktrace1.txt](https://github.com/vaadin/vaadin-gradle-plugin/files/10153376/Stacktrace1.txt)).

There are days without the bug, there are others where every `vaadinBuildFrontend` causes the error.

Identified preconditions: The Bug only surfaces
- when there are changes made to the gradle project containing the Vaadin UI
- when the gradle daemon is used (from the 2nd run)
- ?

**Workaround**
Stop the daemon using `./gradlew --stop` and run the task again.

**To Reproduce**
Reproduction is not possible... it seems to be somewhat dependent on the changes made to the sources.

**Expected behavior**
successful frontend build

**Additional context**

**Comment by skiedrowski** [comment]
There are no native dependencies that I am aware of. TestBench/Selenium is using chromedriver but not as a build dependency, I guess. As you've said, the irregular behaviour does not support this theory. 
Using `--no-daemon` works. But this is just another workaround similar to stopping the daemon which slows down development.
How could I check whether the files are closed properly? #81 is closed.

Thanks for moving to the right repo. 👍

**Comment by Artur-** [comment]
If you have Maven you can run
```
mvn com.github.s4u.plugins:pgpverify-maven-plugin:check
```

to verify all the dependencies used in the project. Not sure if there is something similar for Gradle. It does sound like you have a bad jar file somewhere

**Comment by contradictioned** [comment]
To contribute to the original report,

I experience the same general error. The stack trace looks the same, however I get a different message from gradle:

```
Visiting class [classname] failed with ZipFile invalid LOC header (bad signature).
This might be a broken class in the project.
```

where [classname] is a classname of a class from our own source code.

Stopping the daemon helps me as well.

Desktop: MacOS 13.2, arm, openjdk 19.0.2, gradle 7.6, Vaadin 23.3.4

**Comment by skiedrowski** [comment]
Thanks for chiming in!

In the meantime we've figured the following
- The affected class is from our project (not from a dependency). So the mentioned ZipFile must have been created by gradle (?).
- There are two different classes causing the problem, which are both Kotlin classes (in a mixed Java/Kotlin setup).

Where does the problem originate from? The gradle-vaadin-plugin? Kotlin compiler? Gradle daemon? ...?

**Comment by mpneyer** [comment]
Run into same problem, Java Project (no kotlin) and the problematic class was a generated one using [OpenAPI Generator Plugin](https://plugins.gradle.org/plugin/org.openapi.generator)

> Task :my-project:vaadinBuildFrontend FAILED
> Visiting class ch.my.project.client.some.generated.MyClass failed with ZipFile invalid LOC header (bad signature).
> This might be a broken class in the project.

OS: macOS 14.6
JDK version: 21 (temurin)
Gradle version: 8.8
Vaadin Plugin version: 24.4.4

Workaround: kill gradle daemon

**Comment by marcelbaumann** [comment]
Hello everybody

I run into the same problem multiple times. The problem seems to happen when I run `./gradlew net.tangly.erp.ui:installDist -Pvaadin.productionMode` on the command line and IntelliJ IDEA is open with the same project decide to also start Gradle without the production mode set in the background.

The affected class is one in project e.g. `Visiting class net.tangly.erp.products.ui.ProductsView failed with ZipFile invalid LOC header (bad signature).` It seems to be always a class containing Vaadin code.

(My repository is under `https://github.com/tangly-team/tangly-os`).

The workaround of `./gradlew --stop` works fine. Thanks for the tip @mpneyer 👍 
The workaround is fast and allows us to work daily.

_The problem could be in `vaadinBuildFrontend` running in parallel._

OS: macOS Sonoma 14.6.1
JDK Version: Oracle JDK 21 "22.0.2" 2024-07-16 Java(TM) SE Runtime Environment (build 22.0.2+9-70)
Gradle version: 8.10 (use the Groovy DSL in project)
Vaadin plugin version: 24.4.11

**Comment by deje07** [comment]
It is very disappointing that this bug still exists with latest version. Gradle is clearly not a priority for the team.

**Comment by mcollovati** [comment]
It's quite challenging to provide a proper fix if the issue cannot be reproduced.
If anyone has any ideas on how to reproduce, please let us know.

For example, how the app is launched in the IDE, if there are any specific run configurations used etc

**Comment by mcollovati** [comment]
With the help of Claude, I managed to reproduce this issue (or at least an error that is similar to the reported one). Here are the findings.

## Root cause

`FrontendDependencies.visitClass()` resolves class resources through a `URLClassLoader` created by `ReflectionsClassFinder`. For classes located inside **jar files** (e.g. sibling module jars in a multi-module Gradle project), `url.openStream()` goes through `JarURLConnection` → `JarFileFactory` — a **JVM-global static cache**.

In the Gradle daemon the JVM persists across builds. The `URLClassLoader` created by `ReflectionsClassFinder` is **never explicitly closed**, so its cached `JarFile` handles linger in `JarFileFactory`. When Gradle recompiles a sibling module and rewrites its jar, the stale cached `JarFile` has a central directory pointing to offsets in the **old** file content. Reading at those offsets from the **rewritten** file yields garbage → `ZipException`.

The bug is intermittent because it depends on GC timing: if the old `URLClassLoader` gets garbage-collected between builds, the stale `JarFile` is closed and evicted from the cache, and the next build succeeds.

## Required conditions

- **Multi-module project** (or code-generation plugin producing jars) — user classes must be in jar files on the classpath, not directories
- **Gradle daemon enabled** (from the 2nd run) — `JarFileFactory` static cache must persist across builds
- **Source changes in a sibling module** between runs — triggers jar rewrite, making the cached `JarFile` stale
- **GC has not collected the old `URLClassLoader`** — keeps the stale `JarFile` alive in the cache

This explains why single-module projects seem unaffected: the user's own classes live in `build/classes/java/main/` (a directory, not a jar), so no `ZipFile` is involved.

## Reproducer

I created a reproducer based on the [base-starter-spring-gradle](https://github.com/vaadin/base-starter-spring-gradle) project, converting it into a two-module setup so that a sibling module's jar ends up on the classpath scanned by `vaadinBuildFrontend`.

**Reproducer branch:** https://github.com/vaadin/base-starter-spring-gradle/tree/issues/15458-reproducer (the second commit is to test local fixes)

To run:

```bash
./reproduce.sh
```

The script stops any existing daemon, does an initial clean build to warm up the daemon (populating the `JarFileFactory` cache), then loops: it modifies a source file in the `lib` module and rebuilds. Each iteration rewrites `lib.jar`. If the stale `JarFile` from the previous iteration is still cached, the scan fails.

I was able to reproduce around attempt 11:

```
Visiting class com.example.lib.SharedComponent failed with
  Unexpected end of ZLIB input stream.
  This might be a broken class in the project.
```

There is also a `reproduce-concurrent.sh` script that simulates an IDE recompiling in the background while `vaadinBuildFrontend` runs (matching the scenario described by some reporters in this issue).
Usually reproduce on attempt 4

## Example output 

```
}> ./reproduce.sh

=== Stopping any existing Gradle daemon ===
Stopping Daemon(s)
1 Daemon stopped

=== Initial clean build (warm up daemon, populate JarFile cache) ===
Initial build succeeded.

=== Starting rapid rebuild loop (up to 50 attempts) ===
Each iteration modifies the lib module source, forcing jar rewrite.

Attempt 1: build succeeded (stale cache was likely evicted by GC)
Attempt 2: build succeeded (stale cache was likely evicted by GC)
Attempt 3: build succeeded (stale cache was likely evicted by GC)
Attempt 4: build succeeded (stale cache was likely evicted by GC)
Attempt 5: build succeeded (stale cache was likely evicted by GC)
Attempt 6: build succeeded (stale cache was likely evicted by GC)
Attempt 7: build succeeded (stale cache was likely evicted by GC)
Attempt 8: build succeeded (stale cache was likely evicted by GC)
Attempt 9: build succeeded (stale cache was likely evicted by GC)
Attempt 10: build succeeded (stale cache was likely evicted by GC)
>>> BUG REPRODUCED on attempt 11 <<<

Visiting class com.example.lib.SharedComponent failed with Unexpected end of ZLIB input stream.
This might be a broken class in the project.
```


```
}> ./reproduce-concurrent.sh

=== Stopping any existing Gradle daemon ===
Stopping Daemon(s)
1 Daemon stopped

=== Initial clean build ===
Initial build succeeded.

=== Starting concurrent reproduction ===
A background loop recompiles the lib module every 0.5s.
The foreground runs vaadinBuildFrontend repeatedly.

<-------------> 0% EXECUTING [347ms]
Attempt 2: build succeeded
<-------------> 0% CONFIGURING [62ms]
> root project
Attempt 3: build failed (rc=1), but not the expected ZipException

>>> BUG REPRODUCED on attempt 4 <<<

Visiting class com.example.lib.SharedComponent failed with Unexpected end of ZLIB input stream.
This might be a broken class in the project.
```

**Comment by mcollovati** [comment]
Hopefully what is described above is what you are experiencing.
Now we can try to find a fix for this problem.

**Comment by mcollovati** [comment]
The `ZipException: invalid LOC header (bad signature)` appears to stem from stale JAR file handles surviving across Gradle daemon builds. Based on reading the OpenJDK source, the JVM seems to maintain two independent JAR caching layers, both of which can retain file handles when the same JVM is reused.

The first layer lives inside `URLClassLoader` itself. Each instance caches `JarFile` objects for the URLs it was constructed with, via the internal `URLClassPath` and `JarLoader` classes ([source](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/jdk/internal/loader/URLClassPath.java)). The second layer appears to be a JVM-global static cache in `JarFileFactory` ([source](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/sun/net/www/protocol/jar/JarFileFactory.java)), which is populated whenever code opens a stream on a `jar:` URL. The resulting `JarURLConnection` seems to store the `JarFile` in a static map that outlives any individual class loader. When a sibling module's JAR is rewritten between builds (e.g. by recompilation), these cached handles point to a file whose content no longer matches, producing the `ZipException`.

Closing the `URLClassLoader` releases its own internal JAR cache (layer 1), but this does not appear to affect the JVM-global cache described above. JDK-8246714 ([link](https://bugs.openjdk.org/browse/JDK-8246714)) documents race conditions in `JarFileFactory` that can leave stale entries even after `URLClassLoader.close()` is called. Spring Framework encountered a similar problem and adopted `useCaches(false)` on `jar:` URL connections in their `PathMatchingResourcePatternResolver` ([SPR-4639 / spring-projects/spring-framework#9316](https://github.com/spring-projects/spring-framework/issues/9316)).

Other projects have run into the same class of issue. SpotBugs ([spotbugs/spotbugs#2024](https://github.com/spotbugs/spotbugs/issues/2024)) and CodeNarc ([CodeNarc/CodeNarc#556](https://github.com/CodeNarc/CodeNarc/issues/556)) both fixed their `URLClassLoader` lifecycle but did not address the `JarFileFactory` layer, which left them exposed to stale handles in long-lived processes. The sonar-findbugs plugin ([spotbugs/sonar-findbugs#128](https://github.com/spotbugs/sonar-findbugs/issues/128)) reported a similar stale-JAR problem in its analysis process.

When `getResource()` returns a `jar:` URL and the caller opens a stream on it, the JVM's `JarURLConnection` implementation appears to cache the underlying `JarFile` in a static map managed by `JarFileFactory`. In our case, `FrontendDependencies.visitClass()` calls `classFinder.getResource(...)`, which returns a URL of the form `jar:file:/path/lib.jar!/com/example/Foo.class`. Opening a stream on that URL seems to route through `JarURLConnection` into `JarFileFactory.get()`, which stores the `JarFile` globally. On the next build, `lib.jar` has been rewritten, but the factory still holds the old handle, leading to the `ZipException`.

The potential fix addresses both caching layers in `ReflectionsClassFinder`. For the first layer, `close()` calls `URLClassLoader.close()` to release the class loader's internal JAR cache. For the second layer, `getResource()` could wrap returned `jar:` URLs with a custom `URLStreamHandler` that sets `useCaches(false)` on every connection. This prevents `JarFileFactory` from caching `JarFile` instances in its static map. This approach matches what Spring Framework adopted and avoids any reliance on reflection hacks against JDK internals.

During local testing with only the `URLClassLoader.close()` fix in place, the reproducer was observed to fail around attempt 17 out of 50 iterations. With the `useCaches(false)` fix applied, all 50 attempts passed consistently.

**Comment by mcollovati** [comment]
**Calling for community testers — help us validate a fix for this issue**

Thank you to everyone who has reported and commented here. We know the intermittent `ZipException: ZipFile invalid LOC header (bad signature)` during `vaadinBuildFrontend` has been frustrating, especially because it is difficult to reproduce consistently.

We have a fix candidate ready in #23859, but since we haven't been able to reliably reproduce the original conditions ourselves, we'd like community help validating it before we proceed. Your feedback will directly inform whether this fix is ready to ship.

To keep testing as easy as possible, we'll publish a dedicated pre-release build — no need to build from source. We can provide a version based on **Vaadin 25.1** and potentially **Vaadin 25.0** and **Vaadin 24.9**.

**To volunteer, please:**
1. Reply here with your OS, JDK version, Gradle version, and current Vaadin version
2. We'll publish the appropriate pre-release build. Follow [these instructions](https://vaadin.com/docs/latest/flow/configuration/gradle#pre-release) to configure your project to use Vaadin pre-releases, then update the build files to use `com.vaadin.flow` plugin and swap to the version we provide
3. Use your project as you normally would — if you previously encountered the error, just continue your regular workflow and let us know if it occurs again with the pre-release version
4. Report back here with whether the error occurred or not

If you're still on Vaadin 24.x and the 24.9 build is important to you, please mention that in your reply.

Both "it's fixed" and "still broken" results are equally valuable — thank you in advance for helping!

**Comment by skiedrowski** [comment]
Hi there, I'd like to test the pre-release build.
Environment: mac OS 26, JDK 21.0.x, Gradle 8.14, maybe also Gradle 9.4, Vaadin 24.9.x
... so yes, need a Vaadin 24.9 based pre-release build.

**Comment by mcollovati** [comment]
@skiedrowski We published Flow `24.9.gradle_jarCache_issue_249-SNAPSHOT` to Vaadin pre-releases repository.
To be noted that this is a Flow snapshot; there is no such a version for Vaadin platform.
Therefore, you need to make sure that your project uses the correct version of Flow Gradle plugin instead of Vaadin Gradle plugin.

Disclaimer: I'm not a usual Gradle user, so the following instructions might not be the best, but they work in my case.

**`gradle.properties`** add a separate `flowVersion` property:
```properties
vaadinVersion=24.9.13
flowVersion=24.9.gradle_jarCache_issue_249-SNAPSHOT
```

**`settings.gradle`**  use flowVersion for the plugin and add a resolutionStrategy to map the plugin ID to the actual artifact (SNAPSHOT branches don't publish the Gradle plugin marker artifact):
```grovy
pluginManagement {
    repositories {
        maven { url = 'https://maven.vaadin.com/vaadin-prereleases' }
        gradlePluginPortal()
    }
    plugins {
        id 'com.vaadin.flow' version "${flowVersion}"
    }
    resolutionStrategy {
        eachPlugin {
            if (requested.id.id == 'com.vaadin.flow') {
                useModule("com.vaadin:flow-gradle-plugin:${flowVersion}")
            }
        }
    }
}
```

**`build.gradle`** use `com.vaadin.flow` instead of `com.vaadin` plugin

```groovy
plugins {
	id 'com.vaadin.flow'
}
```

**Comment by mcollovati** [comment]
`25.0.gradle_jarCache_issue_250-SNAPSHOT` for Vaadin 25.0 is also available.

**Comment by skiedrowski** [comment]
I've been using `com.vaadin:flow-gradle-plugin:24.9.gradle_jarCache_issue_249-SNAPSHOT` for 2.5 days and did not get the `Invalid LOC header` error message at all. I'll keep using it and report it if the problem returns. For now it seems to fix the issue.

**Comment by mcollovati** [comment]
@skiedrowski Thanks for testing and for the update! Let's wait a couple of more days to see how it goes and hopefully have someone else's test as well.

**Comment by mcollovati** [comment]
We got no other feedback for the provided patch, but the fix has been merged and will be backported to Vaadin 25.0 and 24.9/10.
Thanks @skiedrowski for taking the time to test the snapshot.
